pkg/logql/log/fmt.go

```diff
@@ -95,23 +95,35 @@ func init() {
 type LineFormatter struct {
 	*template.Template
 	buf *bytes.Buffer
+
+	requiredLabelNamesMap map[string]struct{}
+	requiredLabelNames    []string
 }
 
 // NewFormatter creates a new log line formatter from a given text template.
 func NewFormatter(tmpl string) (*LineFormatter, error) {
 	t, err := template.New("line").Option("missingkey=zero").Funcs(functionMap).Parse(tmpl)
 	if err != nil {
-		return nil, fmt.Errorf("invalid line template: %s", err)
+		return nil, fmt.Errorf("invalid line template: %w", err)
+	}
+	names := uniqueString(listNodeFields([]parse.Node{t.Root}))
+	namesMap := make(map[string]struct{}, len(names))
```

Comment:
Is this just a hashset? We don't seem to use the values.

Reply:
it is. I need more time to thing about this PR. It does improve but I think we can do better at source I might just split this PR into 2.